crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -374,6 +375,122 @@ impl Persistence {
 
         Ok(solution)
     }
+
+    /// Returns the oldest settlement event for which the accociated auction is
+    /// not yet populated in the database.
+    pub async fn get_settlement_without_auction(
+        &self,
+    ) -> Result<Option<domain::settlement::Event>, error::SettlementEvent> {
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["get_settlement_without_auction"])
+            .start_timer();
+
+        let mut ex = self.postgres.pool.begin().await.map_err(DatabaseError)?;
+        let event = database::settlements::get_settlement_without_auction(&mut ex)
+            .await
+            .map_err(DatabaseError)?
+            .map(|event| {
+                let event = settlement::Event {
+                    inner: eth::Event {
+                        block: u64::try_from(event.block_number)
+                            .map_err(|_| error::SettlementEvent::InvalidEvent)?
+                            .into(),
+                        log_index: u64::try_from(event.log_index)
+                            .map_err(|_| error::SettlementEvent::InvalidEvent)?,
+                    },
+                    transaction: eth::TxId(H256(event.tx_hash.0)),
+                };
+                Ok::<_, error::SettlementEvent>(event)
+            })
+            .transpose()?;
+        Ok(event)
+    }
+
+    pub async fn save_settlement(
+        &self,
+        event: domain::settlement::Event,
+        auction_id: domain::auction::Id,
+        settlement: Option<&domain::settlement::Settlement>,
+    ) -> Result<(), error::Settlement> {
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["save_settlement"])
+            .start_timer();
+
+        let mut ex = self.postgres.pool.begin().await.map_err(DatabaseError)?;
+
+        let block_number =
+            i64::try_from(event.inner.block.0).map_err(|_| error::Settlement::InvalidEvent)?;
+        let log_index =
+            i64::try_from(event.inner.log_index).map_err(|_| error::Settlement::InvalidEvent)?;
+
+        database::settlements::update_settlement_auction(
+            &mut ex,
+            block_number,
+            log_index,
+            auction_id,
+        )
+        .await
+        .map_err(DatabaseError)?;
+
+        if let Some(settlement) = settlement {
+            let gas = settlement.gas();
+            let gas_price = settlement.gas_price();
+            let surplus = settlement.native_surplus();
+            let fee = settlement.native_fee();
+            let order_fees = settlement.order_fees();
+
+            tracing::debug!(
```

Comment:
The standard practice is to not embed the values into the format string but rather log all the variables like you did with `auction_id` and `hash` here.

Reply:
fixed